change  parse_x509_cert()  to handle extensions correctly.

Jakuje opened the pull request:
The code attempted to handle extensions assuming extensions were ordered. The
only extension it handled was crl's, but the handling was wrong and I didn't
find any actual use of the crl code. I've changed it to cache all the extensions
and then provided accessors functions to read a specific extension. I needed this
to read the key Usage, but the extension fetching code can work with any extension
(though the caller will need to parse the result. I also added code that parses DN
and returns a specifically requested DN component. I needed this to get the Common
Name for the certificate Subject. This gives the token a 'unique' name rather than
some generic name (like CAC-I or CAC-II). Both of these can be used to enhance the
piv support as well.

(Chunk of #841 providing this functionality to other card)

dengert commented:
OpenSC uses "struct sc_object_id" to store and compare OIDs. Can you change your extension code to  use "const struct sc_object_id *oid" rather "const u8 *type" This is more inline with the rest of the OpenSC handling of OIDs. 

This would require lines like these  to be changed:

u8 *val, *oid;
{ "extensionType", SC_ASN1_OCTET_STRING, SC_ASN1_TAG_OBJECT, SC_ASN1_ALLOC, &oid, &oid_len },

to something like:
struct sc_object_id  oid;
{ "x509v3 entry OID",       SC_ASN1_OBJECT,  SC_ASN1_TAG_OBJECT , 0, &oid, 0 },

and then use the sc_compare_oid to compare. 


Can you change the sc_pkcs15_get_extension routine to accept a "struct sc_object_id" rather then the type?

Jakuje commented:
@dengert I rewrote the arguments to the `struct sc_object_id` and also rewrote the #852 to use this new interface. I hope I got the idea you explained in the previous comment right.
So far tested with cac emulator and it worked fine.

dengert commented on src/libopensc/pkcs15-cert.c:
```diff
@@ -141,6 +133,195 @@ parse_x509_cert(sc_context_t *ctx, struct sc_pkcs15_der *der, struct sc_pkcs15_c
+ *   L: type[] = {0x55, 0x04, 0x07} type_len = 3;
+ *   S: type[] = {0x55, 0x04, 0x08} type_len = 3;
+ *   O: type[] = {0x55, 0x04, 0x0a} type_len = 3;
+ *   OU: type[] = {0x55, 0x04, 0x0b} type_len = 3;
```
The comments should be changed to use sc_object_id way to list a OID.  
For example:
  struct sc_object_id keyUsage_oid={{2,5,29,15,-1}};

dengert commented on src/libopensc/pkcs15-cert.c:
```diff
@@ -141,6 +133,195 @@ parse_x509_cert(sc_context_t *ctx, struct sc_pkcs15_der *der, struct sc_pkcs15_c
+ * if it is not. The data in the extension is extension specific. The following are
+ * common extension values:
+ *   Subject Key ID: type[] = {0x55, 0x1d, 0x0e} type_len = 3;
+ *   Key Usage: type[] = {0x55, 0x1d, 0x0f} type_len = 3;
```
struct sc_object_id keyUsage_oid={{2,5,29,15,-1}};

Jakuje replied:
Makes sense, same as above. Updated now.

dengert commented on src/libopensc/pkcs15-cert.c:
```diff
@@ -141,6 +133,195 @@ parse_x509_cert(sc_context_t *ctx, struct sc_pkcs15_der *der, struct sc_pkcs15_c
+
+
+		/* is it the RN we are looking for */
+		if(sc_compare_oid(&oid, type) == 0) {
```
sc_comapre_id returns 1 if a match and 0 if not. 
Line 274 should be: 
if(sc_compare_oid(&oid, type) != 0) {


I have not checked any other uses of sc_compare_id in this PR.

Jakuje replied:
Thanks. You are right. I read the docs too fast and the wording in comments confused me:

>  * @return 1 if the oids are equal and a non-zero value otherwise

It should probably say

>  * @return 1 if the oids are equal and a zero value otherwise

Will fix in the next commit

Jakuje commented:
@dengert  Thanks for comments. Should be fixed now.
Though some unit tests would be probably something nice-to-have, just to make sure it does what it is supposed to do. Any plans on this?

dengert commented:
I see you have already committed the compare bug. See #847. I want to get @titanous involved, as he requested the change that depends on the #852.

As for testing, #852 I have been using https://github.com/dengert/OpenSC/commit/740a41b  the pkcs15-cert.c part that already parses all the certificates, to look at the keyUsage flags retrieved. I have been using NIST demo cards. Really need @titanous to test out if #852 and https://github.com/dengert/OpenSC/commit/740a41b  do what he is looking for.
